lib/Service/ClientService.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace OCA\Notifications\Service;
+
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\Diagnostics\IQuery;
+use OCP\IDBConnection;
+use OCP\IRequest;
+
+class ClientService {
+	public const DESKTOP_CLIENT_TIMEOUT = 120;
+
+	public function __construct(
+		protected IDBConnection $connection,
+		protected IRequest $request,
+	) {
+	}
+
+	public function hasTalkDesktop(string $userId, int $maxAge = 0): bool {
+		$query = $this->connection->getQueryBuilder();
+		$query->select('name')
+			->from('authtoken')
+			->where($query->expr()->eq('uid', $query->createNamedParameter($userId)));
+
+		if ($maxAge !== 0) {
+			$query->andWhere($query->expr()->gte(
+				'last_activity',
+				$query->createNamedParameter($maxAge, IQueryBuilder::PARAM_INT)
+			));
+		}
+
+		$result = $query->executeQuery();
+		while ($row = $result->fetch()) {
+			\OC::$server->getLogger()->error(json_encode($row['name']));
+			if (preg_match('/ \(Talk Desktop Client - [A-Za-z ]+\)$/', $row['name'])) {
+				$result->closeCursor();
+				return true;
+			}
+		}
+		$result->closeCursor();
```

Comment:
From a perf perspective, wouldn't it be better to add this as a part of SQL query?
Aka `AND name LIKE ' %(Talk Desktop Client -%)'

Reply:
That breaks index usage as far as I remember, and the result set afterwards is so small we are better off checking the few for your user afterwards.
But will check again which indexes are available, but I'm pretty sure name had none